manim/animation/creation.py

```diff
@@ -105,7 +105,7 @@ class ShowPartial(Animation):
 
     """
 
-    def __init__(self, mobject: typing.Union[VMobject, OpenGLVMobject], **kwargs):
+    def __init__(self, mobject: Union[Mobject, OpenGLVMobject, None], **kwargs):
```

Comment:
because the function should take a mobject argument otherwise it is not an init function of an Animation object.

On derived classes, you are allowed to broaden the accepted inputs and to narrow the outputs, but not the other way around.

Reply:
Fair point on the arguments for type-safe vs. not type-safe... but Python is not a strongly typed language (i.e. it uses duck-typing) so I'm unsure if this necessary holds here, or LSP in general for that matter. Here's a [wiki article/section](https://en.wikipedia.org/wiki/Covariance_and_contravariance_(computer_science)#Covariant_method_parameter_type) where the Eiffel language outright rejects LSP.

> The cat shelter illustrates a common phenomenon: it is a kind of animal shelter but has additional restrictions, and it seems reasonable to use inheritance and restricted parameter types to model this.

In either case, this is non-blocking and I'm fine merging as is.

Reply:
To clear up some terminology: Python *is* strongly typed, as opposed to javascript which is weakly typed (you can do things like `'0' == 0`). Rather, python is *dynamically* typed as opposed to statically typed.

I think type annotations should accurately reflect what the code does because the code is the final arbiter of functionality; it's what actually runs. If the code errors if a `Mobject` is passed and requires a `VMobject`, then the annotation should reflect that, otherwise it's a lie. Derived classes that override `__init__` fully replace that method; in code they can add arguments, remove arguments, do whatever, because it's a different method altogether.

Reply:
Thanks for clearing that up @friedkeenan

The docstring section `Raises` right above this code already mentions the precondition/requirement and from what I've briefly read about Python/type hints [in PEP 484](https://www.python.org/dev/peps/pep-0484/#exceptions), exceptions aren't in the scope of type hints.

I believe this is an accurate type hint in that it accepts Mobject and errors if it's not VMobject, or as you've sorta said:
> the code errors if a Mobject is passed and requires a VMobject

so again, LGTM!

Reply:
What you linked seems to be about annotating what exceptions can occur. And technically, you can pass *anything* to a function, the issue is whether it will work or not. Even when type annotations are just suggestions, you're still technically suggesting for the user to pass something that will end up not working.

Ultimately though I don't really care all that strongly about this particular case, but I do find it odd

Reply:
> What you linked seems to be about annotating what exceptions can occur.

I see... so would this be `def __init__(...) -> Optional[Exception]`?

> Even when type annotations are just suggestions, you're still technically suggesting for the user to pass something that will end up not working.
>
>Ultimately though I don't really care all that strongly about this particular case, but I do find it odd

I fully agree (its why I really don't want to block this PR), it's strange to suggest the user pass something that won't work. That's why I brought this up in the first place, but @cobordism made a good argument for LSP/contravariance in the derived class (ShowPartial) whereas I was suggesting ignoring LSP and doing as Eiffel does with covariant parameters.

Reply:
The contravariance makes sense for function arguments, but occasionally it feels strange for `__init__` functions.

It goes like this:
`Transform` is a special type of `Animation`.
Therefore the method `Transform.f(a,b,c)` is an example of a method `Animation.f(x,y,z)` and therefore we cannot put more stringent requirements on a,b,c as we put on x,y,z. 

Slogan: If I can scale a Shape by a float, then you cannot force me to only scale Triangles by ints because a Triangle is a Shape.

For the `__init__` method this view can get lost. Furthermore, what we have is a parallel derivation structure. Here it is in cartoon form:
We have Mobjects and we have a special type of VMobject.
We have Animations and we have a special type of VAnimation.
An Animation takes a Mobject argument and a VAnimation takes a VMobject .. and thus we have broken contravariance.

Is this a sign that maybe our abstractions are "wrong"? Perhaps Animations are a type of Mobject and VAnimations are a type of VMobject? Or maybe there should be an abstract animation, and derived from it a VMobject Animation on the one hand and a Mobject Animation on the other? This is not really a suggestion, just food for thought.